### PR TITLE
Implement status updates for remote conversations

### PR DESCRIPTION
**Pull request by pcapriotti**
This PR implements the ability to update status flags (hidden/muted/archived) for remote conversations.

This adds fields to the cassandra table containing remote conversation membership information for local users, namely `otr_muted_status`, `otr_muted_ref`, `otr_archived`, `otr_archived_ref`, `hidden`, `hidden_ref`. Note that the conversation role is not part of this table, because it is stored in the same backend as the conversation itself.

The qualified endpoint stub for setting self member status has now been fully implemented.

The PR also contains a few refactorings, which were useful because of the split between self member status fields and conversation role:

 - make the user id unqualified in the `get-conversations` RPC (SQCORE-917)
 - extract metadata fields out of the `Conversation` type
 - extract member status fields out of the `LocalMember` type and remove the type variable
 - make the `get-conversations` RPC return a (new type) `RemoteConversation`,
   which does not contain any self member status fields, only a conversation role.

## Checklist

 - [x] The **PR Title** explains the impact of the change.
 - [x] The **PR description** provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] If a cassandra schema migration has been added, I ran **`make git-add-cassandra-schema`** to update the cassandra schema documentation.
 - [x] **changelog.d** contains the following bits of information:
   - [x] A file with the changelog entry in one or more suitable sub-sections. The sub-sections are marked by directories inside `changelog.d`.

### CODE REVIEW

File: services/galley/test/integration/API/Federation.hs [hunk 1]
```diff
@@ -73,45 +74,55 @@ tests s =
...
 getConversationsAllFound = do
-  -- FUTUREWORK: make alice / bob remote users
   [alice, bob] <- randomUsers 2
-  connectUsers alice (singleton bob)
-  -- create & get one2one conv
-  cnv1 <- responseJsonUnsafeWithMsg "conversation" <$> postO2OConv alice bob (Just "gossip1")
-  getConvs alice (Just $ Left [qUnqualified . cnvQualifiedId $ cnv1]) Nothing !!! do
-    const 200 === statusCode
-    const (Just [cnvQualifiedId cnv1]) === fmap (map cnvQualifiedId . convList) . responseJsonUnsafe
+  let aliceQ = Qualified alice (Domain "far-away.example.com")
```
Comment (mdimjasevic) [inline services/galley/test/integration/API/Federation.hs:78]: I don't fully understand what's going on. On line 77 you generate Alice as a random local user and here you associate her with a specific domain that might or might not be the local domain. Perhaps Alice should be generated with `randomId` instead?

File: services/galley/test/integration/API/Federation.hs [hunk 2]
```diff
@@ -73,45 +74,55 @@ tests s =
...
+  carlQ <- Qualified <$> randomUser <*> pure localDomain
+  connectUsers bob (singleton (qUnqualified carlQ))
+
+  putStrLn $ "alice: " <> show (qUnqualified aliceQ)
+  putStrLn $ "bob: " <> show bob
+  putStrLn $ "carl: " <> show (qUnqualified carlQ)
+
+  cnv2 <-
+    responseJsonError
+      =<< postConvWithRemoteUser (qDomain aliceQ) (mkProfile aliceQ (Name "alice")) bob [aliceQ, carlQ]
```
Comment (mdimjasevic) [inline services/galley/test/integration/API/Federation.hs:91]: So, `alice` is no the local domain, but `aliceQ` is on `Domain "far-away.example.com"`, which might be different. Which domain of these two should go as the first argument to `postConvWithRemoteUser`?

File: services/galley/test/integration/API/Federation.hs [hunk 3]
```diff
@@ -73,45 +74,55 @@ tests s =
...
-  localDomain <- viewFederationDomain
-  let aliceQualified = Qualified alice localDomain
   GetConversationsResponse cs <-
     FedGalley.getConversations
       fedGalleyClient
-      (GetConversationsRequest aliceQualified $ qUnqualified . cnvQualifiedId <$> [cnv1, cnv2])
-  let c1 = find ((== cnvQualifiedId cnv1) . cnvQualifiedId) cs
-  let c2 = find ((== cnvQualifiedId cnv2) . cnvQualifiedId) cs
-  liftIO . forM_ [(cnv1, c1), (cnv2, c2)] $ \(expected, actual) -> do
+      (qDomain aliceQ)
```
Comment (mdimjasevic) [inline services/galley/test/integration/API/Federation.hs:106]: The local domain or `Domain "far-away.example.com"` is the intended domain here?

File: services/galley/test/integration/API/Federation.hs [hunk 4]
```diff
@@ -73,45 +74,55 @@ tests s =
...
+      (cnvmName . rcnvMetadata <$> c2)
     assertEqual
-      "self member mismatch"
-      (Just . cmSelf $ cnvMembers expected)
-      (cmSelf . cnvMembers <$> actual)
+      "self member role mismatch"
+      (Just . memConvRoleName . cmSelf $ cnvMembers cnv2)
+      (rcmSelfRole . rcnvMembers <$> c2)
+    putStrLn $ "actual members " <> show (fmap (rcmOthers . rcnvMembers) c2)
+    putStrLn $ "expected members " <> show (cmOthers (cnvMembers cnv2))
```
Comment (mdimjasevic) [inline services/galley/test/integration/API/Federation.hs:121]:
```suggestion
```

File: services/galley/test/integration/API.hs [hunk 5]
```diff
@@ -1906,7 +1920,7 @@ testGetQualifiedRemoteConv = do
   conv <- responseJsonUnsafe <$> (pure respAll <!! const 200 === statusCode)
   -- FUTUREWORK: The backend should augment returned conversation data with
   -- Alice's membership data stored locally
```
Comment (mdimjasevic) [inline services/galley/test/integration/API.hs:1922]: Should this futurework note go?

**Review by mdimjasevic — APPROVED** [review]
This looks great!

I left some comments inlined. I wasn't sure what's going on there so it'd be great if you could clarify that.

File: services/galley/src/Galley/Data.hs [hunk 6]
```diff
@@ -943,6 +999,20 @@ instance IsMemberUpdate OtherMemberUpdate where
           misConvRoleName = omuConvRoleName omu
         }
 
+  -- FUTUREWORK: https://wearezeta.atlassian.net/browse/SQCORE-887
```
[the remark this answers is not in the document]
Comment (pcapriotti) [reply services/galley/src/Galley/Data.hs:1002]: That function will be used to set the conversation role of a local user in a remote conversation, which is currently not supported. The linked issue is exactly about that.

File: services/galley/src/Galley/Data.hs [hunk 7]
```diff
@@ -1050,24 +1137,16 @@ toMember ::
...
-              memConvRoleName = fromMaybe roleNameWireAdmin crn
-            }
+  Maybe LocalMember
+toMember (usr, srv, prv, Just 0, omus, omur, oar, oarr, hid, hidr, crn) =
+  Just $
+    LocalMember
+      { lmId = usr,
+        lmService = newServiceRef <$> srv <*> prv,
+        lmStatus = toMemberStatus (omus, omur, oar, oarr, hid, hidr),
+        lmConvRoleName = fromMaybe roleNameWireAdmin crn
```
[the remark this answers is not in the document]
Comment (pcapriotti) [reply services/galley/src/Galley/Data.hs:1147]: Not sure, but it was already like this. As long as the role is always written in the database, the default here is never used.

File: services/galley/test/unit/Test/Galley/Mapping.hs [hunk 8]
```diff
@@ -23,184 +23,139 @@ module Test.Galley.Mapping where
...
-      I.memOtrArchived = False,
-      I.memOtrArchivedRef = Nothing,
-      I.memHidden = False,
-      I.memHiddenRef = Nothing,
-      I.memConvRoleName = roleNameWireAdmin
-    }
+cnvUids :: Domain -> Conversation -> [Qualified UserId]
+cnvUids dom c =
+  let mems = cnvMembers c
+   in Qualified (memId (cmSelf mems)) dom :
```
[the remark this answers is not in the document]
Comment (pcapriotti) [reply services/galley/test/unit/Test/Galley/Mapping.hs:98]: No `memId` is a field of the `Member` type, which is unchanged. This works with `Wire.API.Conversation`, which contains `Member` values, unlike `Data.Conversation`, which contains `LocalMember` (previously `InternalMember UserId`).

File: services/galley/test/unit/Test/Galley/Mapping.hs [hunk 9]
```diff
@@ -23,184 +23,139 @@ module Test.Galley.Mapping where
...
+  map ((`Qualified` dom) . lmId) (Data.convLocalMembers c)
+    <> map (unTagged . rmId) (Data.convRemoteMembers c)
+
+genLocalMember :: Gen LocalMember
+genLocalMember =
+  LocalMember
+    <$> arbitrary
+    <*> pure defMemberStatus
+    <*> pure Nothing
+    <*> pure roleNameWireAdmin
```
[the remark this answers is not in the document]
Comment (pcapriotti) [reply services/galley/test/unit/Test/Galley/Mapping.hs:112]: It seems to me that roles are mostly irrelevant to this function, but yes, we can also generate them randomly.

File: services/galley/test/integration/API.hs (hunk 5 above)
Comment (pcapriotti) [reply services/galley/test/integration/API.hs]: Yes, done in fe55625f3a10ecd600b47e35cf90042df9403857.

File: services/galley/test/integration/API/Federation.hs (hunk 3 above)
Comment (pcapriotti) [reply services/galley/test/integration/API/Federation.hs:101]: Ditto.

File: services/galley/test/integration/API/Federation.hs (hunk 2 above)
Comment (pcapriotti) [reply services/galley/test/integration/API/Federation.hs:86]: The remote one.

File: services/galley/test/integration/API/Federation.hs (hunk 1 above)
Comment (pcapriotti) [reply services/galley/test/integration/API/Federation.hs]: Yes, that was a mistake. Fixed in 77b62c13cf8022b87afcadc92de82d4154a7a352.

File: services/galley/test/integration/API/Federation.hs [hunk 10]
```diff
@@ -73,45 +74,55 @@ tests s =
...
-    const (Just [cnvQualifiedId cnv1]) === fmap (map cnvQualifiedId . convList) . responseJsonUnsafe
+  let aliceQ = Qualified alice (Domain "far-away.example.com")
+
   -- create & get group conv
-  carl <- randomUser
-  connectUsers alice (singleton carl)
-  cnv2 <- responseJsonUnsafeWithMsg "conversation" <$> postConv alice [bob, carl] (Just "gossip2") [] Nothing Nothing
-  getConvs alice (Just $ Left [qUnqualified . cnvQualifiedId $ cnv2]) Nothing !!! do
+  localDomain <- viewFederationDomain
+  carlQ <- Qualified <$> randomUser <*> pure localDomain
```
[the remark this answers is not in the document]
Comment (pcapriotti) [reply services/galley/test/integration/API/Federation.hs]: Done in 9cd866088f01f24a4629a9e083c0f00d8f8343d4.

File: services/galley/test/unit/Test/Galley/Mapping.hs (hunk 9 above)
[the remark this answers is not in the document]
Comment (pcapriotti) [reply services/galley/test/unit/Test/Galley/Mapping.hs]: Done in b74752ebc1fb7c1b4d776245233c4d95bae217f7.

File: libs/wire-api/src/Wire/API/Conversation.hs [hunk 11]
```diff
@@ -105,62 +117,125 @@ import Wire.API.Conversation.Role (RoleName, roleNameWireAdmin)
...
+      .= lax
+        ( fieldWithDocModifier
+            "message_timer"
+            (description ?~ "Per-conversation message timer (can be null)")
+            (optWithDefault A.Null schema)
+        )
+    <*> cnvmReceiptMode .= lax (field "receipt_mode" (optWithDefault A.Null schema))
+
+instance ToSchema ConversationMetadata where
+  schema = object "ConversationMetadata" conversationMetadataObjectSchema
```
[the remark this answers is not in the document]
Comment (pcapriotti) [reply libs/wire-api/src/Wire/API/Conversation.hs]: Done in a0df7f6ceabb7003a798d80212fde68749e6e806.

**Pull request merged**